### PR TITLE
Add admission webhooks that provision owner annotation and deny to change it

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
It's an initial implementation for admission webhooks that provision owner annotation and deny to change it:
- when workspace is created, user info is provisioned, if a user specifies it - it's overridden
- when workspace is updated:
  - the creator is missing - provision from previous workspace state
       - if the previous state miss it as well - provision the current user as creator
  - the creator is unchanged - do nothing
  - the creator is changed - deny operation with a message, like: `annotation org.eclipse.che.workspace/creator is immutable and must have value: "7c922737-54b6-11ea-bd4d-080027a7b4ae`

What should be considered to be improved:
- It's needed to investigate how we can get certificates for webhook server with OLM, maybe controller should generate certs by itself, for example with https://github.com/operator-framework/operator-sdk/blob/master/pkg/tls/tls.go#L123;
- Revise if it's a good idea to create admission configuration from code, like if we rename configuration name - then in the controller code we should clean up old one, maybe there is a better approach;
- Now webhooks are optional and they will be not configured if there are some issues with K8s cluster or controller configuration. It would be safer to make them required by default and fail if there are some issues but add an ability to disable them with ConfigMap.
- there is no user info on minikube, it's needed to revise what to do? Not Webhooks Configuration just provisions empty value and deny to change it.

Next steps:
1. Make Workspace Controller propagate creator annotations to Workspace Deployment
2. Add ValidaitngWebhook that deny changing creator annotation for workspace-related Deployment/Pod
3. Add ValidatingWebhook that denies creating exec into Pods for everyone except the creator.

As an alternative, we could go with only third webhook that which would find workspace related to current pod with reference Pod->Deployment->Workspace CR, but it means that each exec requires a few additional requests on Webhook Server side...
So, these ways should be compared.

**Minor changes:**
- Added instructions how to remove workspace controller from cluster
- Fix NPE during provisioning userID on k8s cluster
Screenshot_20200226_090814

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/16056

### Is it tested? How?
It's tested on minishift with the following scenario:
1. `minishift addon apply admissions-webhook`. Without it controller will report them webhooks are configured but they won't work. See https://github.com/minishift/minishift/issues/3422
2. Deploy Workspace Controller as usual.
3. Create test workspace and make sure that creator annotation is provisioned, like `kc get workspace cloud-shell -o=yaml | grep creator`.
4. Try to modify creator annotation, like `kc patch workspace cloud-shell --type='merge' -p '{"metadata":{"annotations":{"org.eclipse.che.workspace/creator":"newuser"}}}'`
I'm going to retest it on minikube as well. But it does not prevent from reviewing.

:warning: If you face the following issue:
> Error from server (InternalError): error when creating "./samples/cloud-shell.yaml": Internal error occurred: failed calling admission webhook "mutate-workspaces.che-workspace-controller.svc": Post https://workspace-controller.che-workspace-controller.svc:443/mutate-workspaces?timeout=30s: x509: certificate has expired or is not yet valid

Then probably your minishift has outdated time, to make it up to date, execute:
```
minishift ssh "sudo yum install chrony -y"
  minishift ssh "sudo sed -i 's/makestep 1.0 3/makestep 1.0 -1/g' /etc/chrony.conf"
  minishift ssh "sudo sed -i 's/\#allow/allow/g' /etc/chrony.conf"
  minishift ssh "sudo systemctl enable chronyd"
  minishift ssh "sudo systemctl start chronyd"
  minishift ssh "sudo systemctl status chronyd"
```

Dockerimage to test: `sleshchenko/che-workspace-controller:webhook`

It's also tested on minikube, but user id, in that case, is just an empty string.

### CODE REVIEW

File: pkg/webhook/creator/validatingwebhookcfg.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+//
+package creator
+
+import (
+	"github.com/che-incubator/che-workspace-operator/pkg/webhook/server"
+	"k8s.io/api/admissionregistration/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
```
Comment (amisevsk) [inline pkg/webhook/creator/validatingwebhookcfg.go]:
```suggestion
const (
```

File: README.md [hunk 2]
```diff
@@ -17,8 +17,9 @@ docker build -t quay.io/che-incubator/che-workspace-controller:7.1.0 -f ./build/
 2. `kubectl create namespace che-workspace-controller`
 3. Make sure that the right domain is set in `./deploy/controller_config.yaml` and `./deploy/registry/local/ingress.yaml`
 4. `kubectl apply -f ./deploy/registry/local`
-5. [Optional] Modify ./deploy/controller.yaml and put your docker image and pull policy there.
-6. `kubectl apply -f ./deploy`
+5. Generate certificates for Webhook server by executing: `./deploy/webhook-server-certs/deploy-webhook-server-certs.sh`
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply README.md]: No, when you run controller locally - admission webhooks will be disabled.
There should be some way to let K8s/OS cluster to talk to your local controller but I'm not sure it's a really easy and universal way. So, I just posted pone it if needed - we can return back to this question later

File: deploy/controller.yaml [hunk 3]
```diff
@@ -7,16 +8,16 @@ spec:
...
     metadata:
       labels:
-        name: che-workspace-controller
+        app: che-workspace-controller
     spec:
       serviceAccountName: che-workspace-controller
       containers:
         - name: che-workspace-controller
-          image: quay.io/che-incubator/che-workspace-controller:nightly
+          image: sleshchenko/che-workspace-controller:webhook
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply deploy/controller.yaml]: I dream that all people use my personal images, but since you caught it I must revert :smile:

File: pkg/controller/workspace/utils/utilities.go [hunk 4]
```diff
@@ -10,12 +10,13 @@
 //   Red Hat, Inc. - initial API and implementation
 //
 
-package workspace
+package utils
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/controller/workspace/utils/utilities.go]: `IsOpenShift` & `IsWebhookConfigurationEnabled` are not really workspace utils...
What do you think if we move somewhere in general place, like
`pkg/k8s/client/utils` and name package as k8sutils?

File: pkg/webhook/creator/mutatewebhookcfg.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+//
+package creator
+
+import (
+	"github.com/che-incubator/che-workspace-operator/pkg/webhook/server"
+	"k8s.io/api/admissionregistration/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/webhook/creator/mutatewebhookcfg.go]: Good catch. With var I tried to workaround limitation with const in Go, but it's not good thing to do. So, returned back const

File: pkg/controller/workspace/utils/utilities.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/controller/workspace/utils/utilities.go]: BTW. I've reverted this change since it's not really related to webhook, and we can improve it a separate PR and in some upcoming PR.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply README.md]: :+1:

File: pkg/controller/workspace/utils/utilities.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/controller/workspace/utils/utilities.go]: Sounds good, :+1: for moving generic stuff into another package.

**Comment by davidfestal** [comment]
Great work !

Some general comments:

> It's an initial implementation for admission webhooks that provision owner annotation and deny to change it.
> What should be considered to be improved:
> 
> * It's needed to investigate how we can get certificates for webhook server with OLM, maybe controller should generate certs by itself, for example with https://github.com/operator-framework/operator-sdk/blob/master/pkg/tls/tls.go#L123;

Yes, it seems necessary to have certificates generated by the controller, according to the namespace in which the controller POD has been deployed (namespace is part of the webhook endpoint dns name), and with a CA that is already trusted by the K8S cluster.

> * Revise if it's a good idea to create admission configuration from code, like if we rename configuration name - then in the controller code we should clean up old one, maybe there is a better approach;

In fact to me it seems that the best approach for now is to create the webhook configurations from code (since OLM doesn't allow another option for now). And if you correctly set the owner reference on the created configuration objects, there will be no cleaning problem as soon as you uninstall / reinstall the controller deployment.
This is already done for the metrics service (created by the operator-sdk), and was implemented the same way to managed the service for the embedded plugin registry:
https://github.com/sleshchenko/che-workspace-crd-operator/blob/a99704e63b3d49446a094d2ef6460e383974ac22/pkg/controller/registry/embedded_registry.go#L135  

It seems to me this should be part of this PR

> * Now webhooks are optional and they will be not configured if there are some issues with K8s cluster or controller configuration. It would be safer to make them required by default and fail if there are some issues but add an ability to disable them with ConfigMap.

Fully agree. However before making this reqired by default, we should enable the mechanism for OLM-based deployments

> * there is no user info on minikube, it's needed to revise what to do? Not Webhooks Configuration just provisions empty value and deny to change it.

I would just provision an empty value and deny to change it.

> 
> Next steps:
> 
> 1. Make Workspace Controller propagate creator annotations to Workspace Deployment
> 2. Add ValidaitngWebhook that deny changing creator annotation for workspace-related Deployment/Pod

You might be able to have a single webhook that manages several types of objects no ? If you add several rules in the webhook configuration.

> 3. Add ValidatingWebhook that denies creating exec into Pods for everyone except the creator.
> 
> As an alternative, we could go with only third webhook that which would find workspace related to current pod with reference Pod->Deployment->Workspace CR, but it means that each exec requires a few additional requests on Webhook Server side...
> So, these ways should be compared.

File: pkg/webhook/creator/mutatingwebhookcfg.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	mutateWebhookCfgName       = "mutate-workspace-admission-hooks"
+	mutateWebhookPath          = "/mutate-workspaces"
+	mutateWebhookFailurePolicy = v1beta1.Fail
+)
+
+func buildMutateWebhookCfg() *v1beta1.MutatingWebhookConfiguration {
```
Comment (davidfestal) [inline pkg/webhook/creator/mutatingwebhookcfg.go]: On created webhook configs you have to add the owner reference
If you correctly set the owner reference on the created configuration objects, there will be no cleaning problem as soon as you uninstall / reinstall the controller deployment.
This is already done for the metrics service (created by the operator-sdk), and was implemented the same way to managed the service for the embedded plugin registry:
https://github.com/sleshchenko/che-workspace-crd-operator/blob/a99704e63b3d49446a094d2ef6460e383974ac22/pkg/controller/registry/embedded_registry.go#L135

Additionally, as soon as you have set the owner ref on the webhook configuration you create / update, you would even be able, at controller start (when adding / updating the webhook configs), to detect if you have owned webhook configs with different names than the ones you have to create / update, and remove them.

File: README.md [hunk 7]
```diff
@@ -42,3 +43,14 @@ Running the controller outside of the cluster depends on everything being in one
 1. Follow steps 1-5 for running the controller locally above
 2. `operator-sdk up local --namespace <your namespace> --enable-delve`
 3. Connect debugger to `127.0.0.1:2345` (see config in `.vscode/launch.json`)
+
+### Remove controller from your K8s/OS Cluster
+```sh
+# Delete all workspaces
+kc delete workspaces.workspace.che.eclipse.org --all-namespaces --all
```
Comment (amisevsk) [inline README.md]: Just noticed -- do you also alias `kc=kubectl`? :smile:

Comment (sleshchenko) [reply README.md]: :-D I thought it's default alias ;)
Good catch

File: pkg/controller/workspace/utils/utilities.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/controller/workspace/utils/utilities.go]: `GetContainerNameFromImage` is moved to existing `pkg/controller/modelutils/k8s`

Other like `isOpenShift`, `IsWebhookConfigurationEnabled` are moved into cluster package in `internal/cluster/info.go` file. I'm open for proposal related to better naming.

[the remark this answers is not in the document]
Comment (sleshchenko) [reply pkg/controller/workspace/utils/utilities.go]: P.S. I also discovered that unused imports which are imported with `.` are not detected as such.
So, we need to try to get rid of them, but I've done it yet to avoid huge merge conflicts with your refactoring.

[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/controller/workspace/utils/utilities.go]: Yeah, I try to remove `.` imports when I can but it results in significant changes. We should try to separate that out into one large PR.

File: deploy/webhook-server-certs/deploy-webhook-server-certs.sh [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+PROJECT_FOLDER=$(dirname "$0")/../..
+docker build --no-cache -t generate-webhook-server-certs:latest ${PROJECT_FOLDER}/deploy/webhook-server-certs
+
+TARGET_FOLDER=$PROJECT_FOLDER/build/_output/webhook-certs
+mkdir -p $TARGET_FOLDER
+
+docker run --entrypoint=cat generate-webhook-server-certs:latest /ca/webhook-server-tls.crt > ${TARGET_FOLDER}/webhook-server-tls.crt
+docker run --entrypoint=cat generate-webhook-server-certs:latest /ca/webhook-server-tls.key > ${TARGET_FOLDER}/webhook-server-tls.key
+docker run --entrypoint=cat generate-webhook-server-certs:latest /ca/ca.crt > ${TARGET_FOLDER}/ca.crt
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply deploy/webhook-server-certs/deploy-webhook-server-certs.sh]: fixed in a suggested way. I've started work to make workspace-controller generate TLS certs by itself. So, this script will be removed at all.

**Comment by sleshchenko** [comment]
> I did run into an issue testing on minishift, where the cloud shell workspace would start but I couldn't type anything in the window (Chrome and Firefox), but that may be unrelated to this PR.

@amisevsk Yeah, I face the same issue and believe it's not related to introduced changes. I've tested https://github.com/eclipse/che-machine-exec/pull/82 and it fixes this issue for me. Will finalize it and merge.
